### PR TITLE
Input connection and configuration parameter validation 

### PR DESCRIPTION
**Pull request by mdazam1942**

### CODE REVIEW

File: stix_shifter_modules/qradar/configuration/qradar_config.json [hunk 1]
```diff
@@ -4,7 +4,8 @@
             "default": "QRadar"
         },
         "host": {
-            "type": "text"
+            "type": "text",
+            "regex": "^(([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\\-]*[a-zA-Z0-9])\\.)*([A-Za-z0-9]|[A-Za-z0-9][A-Za-z0-9\\-]*[A-Za-z0-9])$"
```
Comment (yurii-klymenko) [inline stix_shifter_modules/qradar/configuration/qradar_config.json:8]: this regex has to applied to all the hosts.

Comment (yurii-klymenko) [reply stix_shifter_modules/qradar/configuration/qradar_config.json:8]: you need to apply regex to name and description fields as well

File: stix_shifter_modules/qradar/configuration/qradar_config.json [hunk 2]
```diff
@@ -40,6 +41,13 @@
         "selfSignedCert": {
             "type": "password",
             "optional": true
+        },
+        "proxy": {
+            "type": "text",
```
Comment (yurii-klymenko) [inline stix_shifter_modules/qradar/configuration/qradar_config.json:46]: proxy is a section that has to include other fields. https://github.com/opencybersecurityalliance/stix-shifter/blob/develop/stix_shifter_modules/qradar/stix_transmission/arielapiclient.py#L32-L33

File: stix_shifter_utils/utils/param_validator.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+    if input_value <= max_value:
+        return True
+    else:
+        return False
+
+def check_regex(input_value, regex_value):
+    pattern = re.compile(regex_value)
+    match_str = pattern.search(input_value)
+
+    if match_str:
```
Comment (yurii-klymenko) [inline stix_shifter_utils/utils/param_validator.py:137]: return bool(match_str)

File: stix_shifter_utils/utils/param_validator.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+
+def is_leaf(config):
+    if isinstance(config, dict):
+        for key,value in config.items():
+            if isinstance(value, dict):
+                return False
+    return True
+
+def check_min(input_value, min_value):
+    if input_value >= min_value:
```
Comment (yurii-klymenko) [inline stix_shifter_utils/utils/param_validator.py:122]: it is wrong.
max check is wrong as well

File: stix_shifter_utils/utils/param_validator.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+                    validated_params[key] = input_configs[key]
+                    del input_configs[key]
+                else:
+                    if key not in validated_params:
+                        validated_params[key] = dict()
+                    copy_valid_configs(input_configs[key], expected_configs[key], validated_params[key], errors, key_path)
+                    if not input_configs[key]:
+                        del input_configs[key]
+            else:
+                if not 'optional' in expected_configs[key] and not 'default' in expected_configs[key]:
```
Comment (yurii-klymenko) [inline stix_shifter_utils/utils/param_validator.py:107]: expect it to add an error even if there is optional param in options but with 'False' value.
107-111 can be replaced with something like that:
if 'default' in expected_configs[key]:
    validated_params[key] = value['default']
elif not ('optional' in expected_configs[key] or expected_configs[key]['optional']):
    errors.append(key_path)

File: stix_shifter_modules/qradar/configuration/qradar_config.json (hunk 1 above)
Comment (mdazam1942) [reply stix_shifter_modules/qradar/configuration/qradar_config.json:8]: yes I was going to update all config hosts. why do we need to apply regex in name and description? name and description are base config. UI already restricts some characters. and also name and description is not used anywhere in stir-shifter.

File: stix_shifter_modules/qradar/configuration/qradar_config.json (hunk 2 above)
Comment (mdazam1942) [reply stix_shifter_modules/qradar/configuration/qradar_config.json:46]: yes. will update

File: stix_shifter_utils/utils/param_validator.py (hunk 3 above)
Comment (mdazam1942) [reply stix_shifter_utils/utils/param_validator.py:137]: sure

File: stix_shifter_utils/utils/param_validator.py (hunk 5 above)
Comment (yurii-klymenko) [reply stix_shifter_utils/utils/param_validator.py:107]: your conditions validates if there is 'optional' present in the configuration it does not check the value itself.

File: stix_shifter_modules/qradar/configuration/qradar_config.json (hunk 1 above)
Comment (yurii-klymenko) [reply stix_shifter_modules/qradar/configuration/qradar_config.json:8]: it is a security restriction.

File: stix_shifter_utils/utils/param_validator.py (hunk 4 above)
Comment (yurii-klymenko) [reply stix_shifter_utils/utils/param_validator.py:122]: https://github.com/opencybersecurityalliance/stix-shifter/blob/develop/stix_shifter_modules/base_config.json#L7
min/max checking should work to check string length too.

File: stix_shifter_modules/qradar/configuration/qradar_config.json (hunk 1 above)
Comment (mdazam1942) [reply stix_shifter_modules/qradar/configuration/qradar_config.json:8]: @yurii-klymenko I understand the security restriction. But do you know what special characters we should allow for name and description? I asked Luke he doesn't have anything validation for name and description. we need to be align with UI in this case. For simplicity, I can just put a regex that allows only string and integer but not sure that's sufficient.

File: stix_shifter_utils/utils/param_validator.py (hunk 5 above)
Comment (mdazam1942) [reply stix_shifter_utils/utils/param_validator.py:107]: for our use cases, we only specify `optional` when the value is True. I think we had a discussion to remove `"optional": true`. You suggestion may not work. I need to change the conditional check a little bit

File: stix_shifter_utils/utils/param_validator.py (hunk 4 above)
Comment (mdazam1942) [reply stix_shifter_utils/utils/param_validator.py:122]: so the idea is, if `"type": "text"` check length if `"type": "number"` check value

File: stix_shifter_modules/qradar/configuration/qradar_config.json [hunk 6]
```diff
@@ -40,6 +41,28 @@
...
+                "type": "text",
+                "optional": true
+            },
+            "x_forward_proxy_auth": {
+                "type": "password",
+                "optional": true
+            },
+            "optional": true
+        },
+        "data_lake": {
```
[the remark this answers is not in the document]
Comment (mdazam1942) [reply stix_shifter_modules/qradar/configuration/qradar_config.json:64]: I can put boolean type.
`"data_lake": {
            "type": "boolean",
            "default": false
        }`

File: stix_shifter_modules/qradar/configuration/qradar_config.json [hunk 7]
```diff
@@ -40,6 +41,28 @@
...
+            },
+            "x_forward_proxy": {
+                "type": "text",
+                "optional": true
+            },
+            "x_forward_proxy_auth": {
+                "type": "password",
+                "optional": true
+            },
+            "optional": true
```
[the remark this answers is not in the document]
Comment (mdazam1942) [reply stix_shifter_modules/qradar/configuration/qradar_config.json:62]: right. will remove `"optional": true`

File: stix_shifter_utils/utils/param_validator.py [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+        if input_value >= min_value:
+            return True
+        else:
+            return False
+
+    if type == 'number':
+        if len(input_value) >= len(min_value):
+            return True
+        else:
+            return False
```
[the remark this answers is not in the document]
Comment (mdazam1942) [reply stix_shifter_utils/utils/param_validator.py:134]: why do we need min value for password type?

File: stix_shifter_modules/qradar/configuration/qradar_config.json (hunk 7 above)
[the remark this answers is not in the document]
Comment (mdazam1942) [reply stix_shifter_modules/qradar/configuration/qradar_config.json:62]: but that creates another complexity. in current implementation, validator will take it as a required field.

File: stix_shifter_utils/utils/param_validator.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (yurii-klymenko) [reply stix_shifter_utils/utils/param_validator.py:134]: we do not. we want an exception if there is a max param or any other that we do not validate.

File: stix_shifter_modules/qradar/configuration/qradar_config.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (yurii-klymenko) [reply stix_shifter_modules/qradar/configuration/qradar_config.json:64]: it makes more sense.

**Pull request closed without merging**